File: src/main/java/com/technews/aggregate/posts/service/PostsService.java
package com.technews.aggregate.posts.service;

import com.technews.aggregate.posts.constant.PostSubjects;
import com.technews.aggregate.posts.domain.Post;
import com.technews.aggregate.posts.domain.repository.PostsRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.CollectionUtils;

import java.util.List;

@Service
@RequiredArgsConstructor
public class PostsService {

    private final PostsRepository postsRepository;

    @Transactional(readOnly = true)
    public Page<Post> findAllRelease(final PostSubjects subject, final PageRequest pageable, final List<String> categories) {
        return getReleasePage(subject, categories, pageable);
    }

    private Page<Post> getReleasePage(final PostSubjects subject, final List<String> categories, final PageRequest pageable) {
        if (CollectionUtils.isEmpty(categories)) {
            return postsRepository.findBySubject(subject.value(), pageable);
        }

        return postsRepository.findBySubjectAndCategoryIn(subject.value(), categories, pageable);
    }
}

File: src/main/kotlin/com/technews/aggregate/posts/service/PostsService.kt
package com.technews.aggregate.posts.service

import com.technews.aggregate.posts.constant.PostSubjects
import com.technews.aggregate.posts.domain.Post
import com.technews.aggregate.posts.domain.repository.PostsRepository
import com.technews.aggregate.posts.dto.PostResponse
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class PostsService(
    private val postsRepository: PostsRepository,
) {
    @Transactional(readOnly = true)
    fun findAllRelease(
        subject: PostSubjects,
        pageable: PageRequest,
        categories: List<String>? = emptyList(),
    ): Page<PostResponse> {
        val postPage = if (categories.isNullOrEmpty()) {
            postsRepository.findBySubject(subject.value, pageable)
        } else {
            postsRepository.findBySubjectAndCategoryIn(subject.value, categories, pageable)
        }

        return postPage.map { post ->
            PostResponse(
                id = post.id,
                subject = post.subject,
                title = post.title,
                url = post.url,
                category = post.category,
                writer = post.writer,
                date = post.date,
                tags = post.tags,
                shared = post.shared,
                createdDt = post.createdDt,
            )
        }
    }
}
